### PR TITLE
Basic type-checking with mypy and pyright

### PR DESCRIPTION
**Pull request by Avasam**
This is the PR that finally makes basic type-checking validation of public methods possible, easing the addition of 3.7+ type annotations.

In its current state, a lot of checks are disabled, and some areas of code are completely ignored. But a handful of critical issues are now checked by GitHub actions and will help prevent some regressions.
A few typing fixes have already been provided where I considered they would be worth over disabling globally.

Once this PR is merged, I can in parallel start:
- Progressively adding type annotations for public methods
- Progressively adding explicit type annotations for public variables that are not fully inferred (like partially unknown container types)
- Progressively turning back on some stricter checks in mypy, pyright
- Validating typing standards using https://github.com/PyCQA/flake8-pyi or https://beta.ruff.rs/docs/rules/#flake8-pyi-pyi

### CODE REVIEW

File: com/win32com/client/build.py
Comment (Avasam) [inline com/win32com/client/build.py:72]: Is this intended to be public API? If not I'd rename it to `NoTranslateSet`

File: com/win32com/client/gencache.py
Comment (Avasam) [inline com/win32com/client/gencache.py:60]: `Any` should be used sparingly, it's an escape hatch. For instance: When it's impossible to represent a type accurately using the current type system.

In this case, we have no way to represent this type though python code (for now).

**Comment by Avasam** [comment]
I'm not sure I understand the test failure here. Maybe an infinite loop/wait that causes it to run for hours until it times out?
Merging other related PRs first will lead to less changes here and make it easier for me to figure out where the issue comes from.

File: mypy.ini
Comment (Avasam) [inline mypy.ini:36]: Any idea where `CallTipWindow`, `dde` and `pywin32_system32` come from?

Comment (mhammond) [reply mypy.ini:36]: dde is a module - https://github.com/mhammond/pywin32/blob/main/setup.py#L2025
pywin32_system32 is a hack setup.py does to put the DLLs in a place where they aren't technically a package, eg, https://github.com/mhammond/pywin32/blob/main/setup.py#L337-L339
IIRC, CallTipWindow was a part of idle which we hackily reused -  https://github.com/mhammond/pywin32/blob/4025d7680552beb200688911e3ec87e11acf9365/Pythonwin/pywin/idle/CallTips.py#L38-L41 - I doubt it works at all any more.

**Comment by mhammond** [comment]
Thanks, but I'm seeing many annotations in the github "changed files" view which shouldn't be there, near the end of the PR:

image

I think we need to get rid of these before it can merge.

**Comment by mhammond** [comment]
Re "[Solve some pyright warnings in changed files](https://github.com/mhammond/pywin32/pull/2102/commits/49c1bbb010b82605e4c2969d24435042abe23640)" - does this imply that in the future, when other changed files are touched, these same warnings might appear in those files?

**Comment by Avasam** [comment]
> Re "[Solve some pyright warnings in changed files](https://github.com/mhammond/pywin32/pull/2102/commits/49c1bbb010b82605e4c2969d24435042abe23640)" - does this imply that in the future, when other changed files are touched, these same warnings might appear in those files?

If a warning is present in a file that has been modified, yes.
Not that these are warnings, not error. They are meant to say "this needs to be looked at, or fixed eventually (but won't block the PR)". There is currently 471 errors that I have downgraded to warnings in `pyrightconfig.json` because otherwise the CI wouldn't pass in its current state (and the goal of this PR is to do the minimum so that public type-annotations can be type-checked), but they are valid pyright reports.

As for the fact that they show up as "GitHub annotations", I think I can disable that entirely from what I'm reading. https://github.com/actions/toolkit/blob/master/docs/commands.md#problem-matchers

File: pyrightconfig.json
Comment (jakebailey) [inline pyrightconfig.json:4]: You don't want to have this set; this config file overrides flags so the CI python-version will be overridden: https://github.com/mhammond/pywin32/actions/runs/8084830821/job/22091009766?pr=2102#step:3:9

https://github.com/microsoft/pyright/issues/7330

Comment (jakebailey) [reply pyrightconfig.json:4]: I would recommend unsetting this, and then also not sepcifying the version in the action, instead leaving it to the interpreter set up by setup-python.

Comment (Avasam) [reply pyrightconfig.json:4]: Thanks for pointing out this gotcha. It's been re-openned as a feature request in pyright, but I see the value of adding a warning for it in pyright-action.

In the mean time, I'll update the CIs I work on and put a mental pin in that to change back later if the way it works is changed.

Comment (Avasam) [reply pyrightconfig.json]: > you're already getting it implicitly set via the interpreter you're using which is sufficient IMO.

I don't see the harm either in being extra explicit here, especially in the context of that feature request, where pyright works in unexpected ways with its configs, and we would re-introduce it if the feature request is accepted.

File: pyrightconfig.json
Comment (Avasam) [inline pyrightconfig.json:4]:
```suggestion
  "pythonVersion": "3.7",
```
https://github.com/microsoft/pyright/releases/tag/1.1.352

File: pyrightconfig.json
Comment (Avasam) [reply pyrightconfig.json:4]: Eric is always so fast, this is already released: https://github.com/microsoft/pyright/releases/tag/1.1.352

Comment (jakebailey) [reply pyrightconfig.json:4]: The latest action also doesn't warn anymore (for 1.1.352+), so you should be able to bump that as well!

**Comment by Avasam** [comment]
As a sidenote before this comes up: Test dependencies should probably be pinned. And even better: written to a `requirements` file so they can be deduplicated, easy to install, and their download cached by the `setup-python`. But no dependency is pinned currently in pywin32. So that's a change I'd propose in a separate follow-up PR.

**Comment by Avasam** [comment]
@mhammond I think all of your latest concerns have been addressed here.

Btw all of my currently opened PRs are ready for review / hoping for a merge. You can use this query to ignore `adodbapi`-related ones: https://github.com/mhammond/pywin32/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+author%3AAvasam+NOT+adodbapi

**Review by mhammond — APPROVED** [review]
Thanks!

**Pull request merged**
